includes/class-wc-cart-fees.php

```diff
@@ -151,4 +151,15 @@ protected function sort_fees_callback( $a, $b ) {
 	private function generate_id( $fee ) {
 		return sanitize_title( $fee->name );
 	}
+
+	/**
+	 * Remove all fees belonging to a specific cart.
+	 *
+	 * @since 3.3.0
+	 */
+	public function remove_all_cart_fees( $cart ) {
+		if ( $this->cart === $cart ) {
```

Comment:
Hacky. Ideally we'd want to be able to pass context to `WC_Cart` and friends, instead of having them default to session data.

cc @mikejolley

Reply:
> Hacky

yeah, it feels strange and hacking to me too. The other possible fix I considered was to remove the way it's currently being done and do it in the `WC_Cart` class itself. So instead of removing the fees via a hook and callback, do something like this instead:

```diff
diff --git a/includes/class-wc-cart-fees.php b/includes/class-wc-cart-fees.php
index cbe9433..0e9aee3 100644
--- a/includes/class-wc-cart-fees.php
+++ b/includes/class-wc-cart-fees.php
@@ -69,8 +69,6 @@ final class WC_Cart_Fees {
         * Register methods for this object on the appropriate WordPress hooks.
         */
        public function init() {
-               add_action( 'woocommerce_cart_emptied', array( $this, 'remove_all_fees' ) );
-               add_action( 'woocommerce_cart_reset', array( $this, 'remove_all_fees' ) );
        }
 
        /**
diff --git a/includes/class-wc-cart.php b/includes/class-wc-cart.php
index 248079b..f690ad2 100644
--- a/includes/class-wc-cart.php
+++ b/includes/class-wc-cart.php
@@ -640,6 +640,8 @@ class WC_Cart extends WC_Legacy_Cart {
                        $this->session->persistent_cart_destroy();
                }
 
+               $this->fees_api->remove_all_fees();
+
                do_action( 'woocommerce_cart_emptied' );
        }
 
@@ -1964,6 +1966,7 @@ class WC_Cart extends WC_Legacy_Cart {
         */
        private function reset_totals() {
                $this->totals = $this->default_totals;
+               $this->fees_api->remove_all_fees();
                do_action( 'woocommerce_cart_reset', $this, false );
        }
 }
```

That way the fees associated with the cart being reset or emptied are the only ones being removed. Let me know if you would prefer that approach.

Reply:
Sounds like a better option. I'll close this out and let you submit that instead. Thanks!